hil-test/tests/clock_monitor.rs

```diff
@@ -0,0 +1,52 @@
+//! Clock Monitor Test
+
+//% CHIPS: esp32 esp32c2 esp32c3 esp32c6 esp32h2 esp32s2 esp32s3
+
+#![no_std]
+#![no_main]
+
+use defmt_rtt as _;
+use esp_backtrace as _;
+use esp_hal::{clock::ClockControl, peripherals::Peripherals, prelude::*, rtc_cntl::Rtc};
+
+struct Context<'a> {
+    rtc: Rtc<'a>,
+}
+
+impl Context<'_> {
+    pub fn init() -> Self {
+        let peripherals = Peripherals::take();
+        let system = peripherals.SYSTEM.split();
+        ClockControl::boot_defaults(system.clock_control).freeze();
+
+        let rtc = Rtc::new(peripherals.LPWR, None);
+
+        Context { rtc }
+    }
+}
+
+#[cfg(test)]
+#[embedded_test::tests]
+mod tests {
+    use super::*;
+
+    #[init]
+    fn init() -> Context<'static> {
+        Context::init()
+    }
+
+    #[test]
+    fn test_estimated_clock(mut ctx: Context<'static>) {
+        // We call the function twice since sometimes the first call gives a
+        // wrong result
+        ctx.rtc.estimate_xtal_frequency();
```

Comment:
Hmm, this feels like a bit of code smell. Ideally `estimate_xtal_frequency` should handle this case and ensure that it's a reasonable value before returning. What does esp-idf do?

Reply:
Yes, I wanted to change the implementation (for C6/H2 only IIRC) for some time but never got to it. I think that function isn't super useful for C6/H2 anyways since they don't come with different XTAL frequencies but getting very weird results sometimes is not good

Reply:
If this was an example I wouldn't care so much, but test cases should be correct before we rely on them for regression testing :D. So I think its worth spending the time here to ensure we've got this right.

Reply:
In, [61cf62d](https://github.com/esp-rs/esp-hal/pull/1425/commits/61cf62db7783697f8c9d0029c1452e0cc8235126) I added @bjoernQ idea to get the value again if the value its very off.